### PR TITLE
added validation for tasks.js

### PR DESCRIPTION
**Pull request by apoopan**
-Added userId in create for tasks.js
-Did data side and route side validation for tasks
-did xss checks
-checked if passed id belongs to user in session
-Solved some more bug fixes

### CODE REVIEW

File: data/tasks.js [hunk 1]
```diff
@@ -112,37 +141,54 @@ let exportedMethods = {
...
+    async create(userId , name, importance, deadlineDate) {
+    //  console.log("here")
+      userId = validatetId(xss(userId)); 
+
+      const targetUser = await userData.get(userId);
+    
+    name=name.trim()
 
-    async create(name, importance, deadlineDate) {
-      if(!verify.validString(name)){
```
Comment (jinfante708) [inline data/tasks.js:168]: I think you might have deleted the string verification for `name`, please re-add it

File: routes/tasks.js [hunk 2]
```diff
@@ -5,108 +5,285 @@ const router = express.Router();
...
+  name=name.trim()
+
+  if(!verify.validDate(xss(deadlineDate)))
+  {
+    throw "Deadline date is Not a valid date";
+  }
+
+  importance  = parseInt(importance);
+
+  // if(!verify.validNumber(xss(importance)))
```
Comment (jinfante708) [inline routes/tasks.js]: I think we still need to verify importance as a valid number, parseInt returns 'NaN' if its not a number which isnt enough for an error check, you can use the `verify.validFormNumber()` method

File: routes/tasks.js [hunk 3]
```diff
@@ -5,108 +5,285 @@ const router = express.Router();
...
+importance  = parseInt(importance);
+// console.log(importance)
+// console.log(typeof(importance))
+
+// if(typeof(importance)!=number)
+// {
+//   throw "importance not a valid number"
+// }
+
+// if(!verify.validNumber(xss(importance)))
```
Comment (jinfante708) [inline routes/tasks.js]: Same here, we still need to verify that importance is a valid number

File: data/tasks.js [hunk 4]
```diff
@@ -24,51 +25,72 @@ function compare(a, b) {
...
+  //  if(alltasks.length==0)
+  //  {throw "No tasks present for the provided tasklist"}
+
+    PriorityInDescendingorder = alltasks.sort(compare);
+    return PriorityInDescendingorder;
+  },
+
   async gettasklistid(taskid)
-  {
+  {
```
Comment (AnkitKachhadiya) [inline data/tasks.js]: Please do user id check with tasklistid

File: data/tasks.js [hunk 5]
```diff
@@ -24,51 +25,72 @@ function compare(a, b) {
...
-    PriorityInDescendingorder = alltasks.sort(compare);
-    return PriorityInDescendingorder;
   },
 
   
+  
  async get(id){
-  //  console.log(id)
-  id=id.trim()
+  const Id = validatetId(xss(id));
```
Comment (AnkitKachhadiya) [inline data/tasks.js:98]: Please do user id check with id

File: data/tasks.js [hunk 6]
```diff
@@ -24,51 +25,72 @@ function compare(a, b) {
...
+  const task= await tasksCollection.findOne({ _id: Id});
+  if (!task) {
+  throw "task not found"
+  }
   return task
 },
 
-
   async remove(id){
+    const Id = validatetId(xss(id));
```
Comment (AnkitKachhadiya) [inline data/tasks.js:109]: Please do user id check with id

File: data/tasks.js [hunk 7]
```diff
@@ -84,26 +106,33 @@ let exportedMethods = {
...
-       return {"taskID": id, "deleted": true}
+             if (updateInfo.modifiedCount !== 1) {
+              throw  "Could not update"
+            
+            }
+       return {"taskID": Id, "deleted": true}
  },
 
   async complete(id){
+       const Id = validatetId(xss(id));
```
Comment (AnkitKachhadiya) [inline data/tasks.js:143]: Please do user id check with id

**Review by AnkitKachhadiya — CHANGES_REQUESTED** [review]

**Pull request merged**